design-proposals/generated-disk-serial.md

```diff
@@ -0,0 +1,151 @@
+
+The serial number has to be the same across VM restarts and storage migration,
+so it has to be stored somewhere, or generated deterministically.
+
+One possibility is to store it in an annotation on the PVC object.
+The annotation would be created when the VMI is created. 
+Then it will be read by `virt-handler` when creating libvirt domain XML.
+
+### Other persistence options
+There are several other options where to persist the serial number:
+
+- **Deterministic generation**: We can use name and namespace, or UID fields of the PVC
+  to deterministically generate a serial number. This could be an issue for storage migration,
+  if these values change.
+  
+- **Storing the serial in a ConfigMap**: We can create a ConfigMap that would store all disk serials for a VM.
+  Then it would be simple to generate serials for disks that don't use PVC source. However, creating a ConfigMap
+  and keeping it in sync with the VM is additional complexity.
+
+- **Storing the serial inside the disk**: The serial could be stored as a file inside the disk.
+  This would probably only work for filesystem PVCs.
+
+## API Examples
+
+### Existing API to set serial number manually
+
+The serial number can be specified manually in VM spec here:
+```yaml
+apiVersion: kubevirt.io/v1
+kind: VirtualMachine
+metadata:
+  name: example-vm
+spec:
+  template:
+    spec:
+      domain:
+        devices:
+          disks:
+            - name: disk
+              serial: 1234567890 # <-- serial number
+              disk:
+                bus: virtio
+      volumes:
+        - name: disk
+          persistentVolumeClaim:
+            claimName: example-pvc
+```
+
+### New API
+
+To enable the serial generation, new filed `generateDiskSerials` will be added to the VMI `.spec`:
+```yaml
+apiVersion: kubevirt.io/v1
+kind: VirtualMachine
+metadata:
+  name: example-vm
+spec:
+  template:
+    spec:
+      domain:
+        devices:
+          generateDiskSerials: true # <-- new field
+          disks:
+            - name: disk
+              disk:
+                bus: virtio
+      volumes:
+        - name: disk
+          persistentVolumeClaim:
+            claimName: example-pvc
+```
+
+The generated serial will be stored as an annotation in the PVC:
+```yaml
+apiVersion: v1
+kind: PersistentVolumeClaim
+metadata:
+  name: example-pvc
+  annotations:
+    kubevirt.io/disk-serial: 123456789 # <-- new annotation
+spec:
+  accessModes:
+    - ReadWriteMany
+  resources:
+    requests:
+      storage: 10Gi
+```
+
+## Scalability
+Serials are generated during VMI creation, if the annotation does not exist.
+
+## Update/Rollback Compatibility
+The feature is backward compatible. Existing VMs remain unaffected unless the new field is set.
+Rollback ignores the new field.
+
+## Functional Testing Approach
+
+There are already functional tests to check disk serial that
+is specified manually. We can copy these tests and enable `generateDiskSerials`
```

Comment:
What would be the cost of this approach? Would it increase the test suite execution time? Is there a way to avoid longer test suite execution times?

Reply:
I've added text about unit tests.